### ISSUE TITLE
Indexing broken for LQPackedQ for non-square matrices

### ISSUE THREAD

**Issue by andreasnoack** [JuliaLang/julia#23733]
```julia
julia> lqfact(randn(3,4))[:Q][1,1]
ERROR: DimensionMismatch("dot product arguments have lengths 3 and 4")
Stacktrace:
 [1] dot(::Array{Float64,1}, ::Array{Float64,1}) at ./linalg/blas.jl:311
 [2] getindex(::Base.LinAlg.LQPackedQ{Float64,Array{Float64,2}}, ::Int64, ::Int64) at ./linalg/lq.jl:77

julia> lqfact(randn(3,3))[:Q][1,1]
-0.3461637478517863

julia> lqfact(randn(4,3))[:Q][1,1]
ERROR: DimensionMismatch("dot product arguments have lengths 4 and 3")
Stacktrace:
 [1] dot(::Array{Float64,1}, ::Array{Float64,1}) at ./linalg/blas.jl:311
 [2] getindex(::Base.LinAlg.LQPackedQ{Float64,Array{Float64,2}}, ::Int64, ::Int64) at ./linalg/lq.jl:77
```

**Comment by Sacha0** [comment]
Reduced to
```julia
julia> lqfact(randn(3, 4))[:Q] * [1, 0, 0, 0]
4-element Array{Float64,1}:
 -0.6525221125001841
  0.16438982421875137
 -0.7284684272483514
  0.1285481578784342
```
due to the following methods in base/linalg/lq.jl
```julia
A_mul_B!(A::LQPackedQ{T}, B::StridedVecOrMat{T}) where {T<:BlasFloat} =
    LAPACK.ormlq!('L','N',A.factors,A.τ,B)
function (*)(A::LQPackedQ, B::StridedVecOrMat)
    TAB = promote_type(eltype(A), eltype(B))
    A_mul_B!(convert(AbstractMatrix{TAB}, A), copy_oftype(B, TAB))
end
```
with the issue being that `ormlq!` performs the operation for the square form of `Q`?

**Comment by Sacha0** [comment]
Using `x = zeros(eltype, size(A, 2))` in
```
function getindex(A::LQPackedQ, i::Integer, j::Integer)
    x = zeros(eltype(A), size(A, 1))
    x[i] = 1
    y = zeros(eltype(A), size(A, 2))
    y[j] = 1
    return dot(x, A*y)
end
```
should do the trick? Unless `LQPackedQ` should behave like a thin rather than square `Q` in implicit operations like these?

**Comment by andreasnoack** [comment]
That would fix it but it is better just to extract the element, i.e. `(A*y)[j]` or maybe even `A_mul_B!(A, y)[j]` although it is not performance critical.

**Comment by Sacha0** [comment]
Yes, your suggestions seem still better :).

**Comment by andreasnoack** [comment]
Fixed by #23782
